### PR TITLE
pkg/storage/redis: fix multiple data race

### PR DESCRIPTION
**Pull request by cuonglm**
## Summary
There are two data race in current code:

 - Call to close pub sub conn and renew pub sub conn
 - Call to close notify channel and send data to it

Fixing them by:

 - Moving pub sub conn creation/renew in the doNotifyLoop
 - Add a lock to guard before close/send data to channel, and also add
 another check for context was done at the beginning of notify loop.

Verifying by running:

```
	for _ in {1..100}; do
	  go test -race -count=1 ./pkg/storage/redis/...
	done
```

with no failure.

**Checklist**:
- [ ] add related issues
- [ ] updated docs
- [ ] updated unit tests
- [ ] updated UPGRADING.md
- [x] ready for review

### CODE REVIEW

File: pkg/storage/redis/redis.go [hunk 1]
```diff
@@ -257,7 +273,10 @@ func (db *DB) doNotifyLoop(ctx context.Context, ch chan struct{}, psc *redis.Pub
 			case <-ctx.Done():
 				log.Warn().Err(ctx.Err()).Msg("context done, stop receive from redis channel")
 				return
-			case ch <- struct{}{}:
+			default:
+				db.notifyChMu.Lock()
+				ch <- struct{}{}
```
Comment (calebdoxsey) [inline pkg/storage/redis/redis.go:278]: this can block and leave the notify mutex permanently locked

Comment (cuonglm) [reply pkg/storage/redis/redis.go:278]: In theory, yes. In practice, `doNotifyLoop` is the only one that send data to the channel, so it can't block.

Comment (cuonglm) [reply pkg/storage/redis/redis.go:278]: If you worried about that, then I think we can add another select case to ctx.Done(), PR is coming.

Comment (cuonglm) [reply pkg/storage/redis/redis.go:278]: @calebdoxsey fixed.

**Pull request merged**